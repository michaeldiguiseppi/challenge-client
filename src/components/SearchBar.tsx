import React from "react";
import "./stylesheets/SearchBar.css";

export type SearchBarProps = {
  autocomplete: (e: React.ChangeEvent<HTMLInputElement>) => void;
  value: string;
  clearInput: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
};

const SearchBar: React.FC<SearchBarProps> = ({
  autocomplete,
  value,
  clearInput,
}) => (
  <div className="search-bar-container">
    <input
      data-testid="search-bar"
      className="search-bar-field"
      placeholder="Search employees by first or last name..."
      onChange={e => autocomplete(e)}
      value={value}
    />
    <button
      type="reset"
      className="search-bar-clear-button"
      data-testid="search-bar-clear-button"
      onClick={e => clearInput(e)}
    >
      Clear
    </button>
  </div>
);

export default SearchBar;
