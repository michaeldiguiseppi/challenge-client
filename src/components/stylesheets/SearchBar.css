.search-bar-container {
  display: flex;
  justify-content: center;
}

.search-bar-field {
  height: 40px;
  border-radius: 25px;
  margin: 10px;
  width: 80%;
  border: 1px solid #333;
  outline: none;
  text-align: center;
}

.search-bar-clear-button {
  position: relative;
  left: -60px;
  background-color: transparent;
  border: none;
}
