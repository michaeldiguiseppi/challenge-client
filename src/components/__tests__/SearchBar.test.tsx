import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import SearchBar, { SearchBarProps } from "../SearchBar";

beforeEach(cleanup);

const searchBarProps: SearchBarProps = {
  autocomplete: jest.fn(),
  value: "",
  clearInput: jest.fn(),
};

describe("<SearchBar />", () => {
  it("renders without crashing", () => {
    const { container } = render(<SearchBar {...searchBarProps} />);
    expect(container).toMatchSnapshot();
  });

  it("should render a search bar", () => {
    const { getByTestId } = render(<SearchBar {...searchBarProps} />);
    getByTestId(/^search-bar$/);
  });

  it("should call the autocomplete function from props when changing the input value of the search bar", () => {
    const { getByTestId } = render(<SearchBar {...searchBarProps} />);
    const inputEl = getByTestId(/^search-bar$/);
    fireEvent.change(inputEl, { target: { value: "a" } });
    expect(searchBarProps.autocomplete).toHaveBeenCalledTimes(1);
  });

  it("should call the autocomplete function from props the same number of times as characters input to the search bar", () => {
    const { getByTestId } = render(<SearchBar {...searchBarProps} />);
    const inputEl = getByTestId(/^search-bar$/);
    fireEvent.change(inputEl, { target: { value: "a" } });
    fireEvent.change(inputEl, { target: { value: "as" } });
    fireEvent.change(inputEl, { target: { value: "ast" } });
    expect(searchBarProps.autocomplete).toHaveBeenCalledTimes(3);
  });

  it("should render a clear input button", () => {
    const { getByTestId } = render(<SearchBar {...searchBarProps} />);
    getByTestId(/search-bar-clear-button/);
  });

  it("should clear the search bar when the clear button is pressed", () => {
    const { getByTestId } = render(<SearchBar {...searchBarProps} />);
    const searchBarField = getByTestId(/^search-bar$/);
    const clearButton = getByTestId(/search-bar-clear-button/);
    fireEvent.change(searchBarField, { target: { value: "a" } });
    fireEvent.click(clearButton);
    expect(searchBarProps.clearInput).toHaveBeenCalledTimes(1);
  });
});
