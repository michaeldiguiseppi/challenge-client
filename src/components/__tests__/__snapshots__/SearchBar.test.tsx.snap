// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> renders without crashing 1`] = `
<div>
  <div
    class="search-bar-container"
  >
    <input
      class="search-bar-field"
      data-testid="search-bar"
      placeholder="Search employees by first or last name..."
      value=""
    />
    <button
      class="search-bar-clear-button"
      data-testid="search-bar-clear-button"
      type="reset"
    >
      Clear
    </button>
  </div>
</div>
`;
